src/graphics/env-lighting.js

```diff
@@ -0,0 +1,397 @@
+        reprojectTexture(prevLevel || target, level, {
+            numSamples: 1
+        });
+
+        // copy into target's mipmap
+        copyMipmap(target, level);
+
+        if (prevLevel) {
+            prevLevel.destroy();
+        }
+        prevLevel = level;
+    }
+
+    if (prevLevel) {
+        prevLevel.destroy();
+    }
+
+    // set filtering while bypassing the engine
+    device.setTexture(target, 0);
+    gl.texParameteri(target._glTarget, gl.TEXTURE_MAG_FILTER, gl.LINEAR);
+    gl.texParameteri(target._glTarget, gl.TEXTURE_MIN_FILTER, gl.LINEAR_MIPMAP_LINEAR);
+    target._mipmaps = true;
+};
+
+// helper functions to support prefiltering lighting data
+class EnvLighting {
+    /**
+     * @private
+     * @function
+     * @name generateSkyboxCubemap
+     * @description Generate a skybox cubemap in the correct pixel format from the source texture.
+     * @param {Texture} source - The source texture. This is either a 2d texture in equirect format or a cubemap.
+     * @param {number} size - Size of the resulting texture. Specify 0 for automatic sizing.
+     * @returns {Texture} The resulting cubemap.
+     */
+    static generateSkyboxCubemap(source, size = 0) {
+        if (size === 0) {
+            size = source.cubemap ? source.width : source.width / 4;
+        }
+
+        const device = source.device;
+
+        // generate faces cubemap
+        const result = new Texture(device, {
+            name: 'skyboxFaces',
+            cubemap: true,
+            width: size,
+            height: size,
+            type: TEXTURETYPE_RGBM,
+            addressU: ADDRESS_CLAMP_TO_EDGE,
+            addressV: ADDRESS_CLAMP_TO_EDGE,
+            fixCubemapSeams: fixCubemapSeams,
+            mipmaps: false
+        });
+
+        reprojectTexture(source, result, {
+            numSamples: 1024
+        });
+
+        return result;
+    }
+
+    /**
+     * @private
+     * @function
+     * @name generateLightingSource
+     * @description Create a texture in the format needed to precalculate lighting data.
+     * @param {Texture} source - The source texture. This is either a 2d texture in equirect format or a cubemap.
+     * @returns {Texture} The resulting cubemap.
+     */
+    static generateLightingSource(source) {
+        const device = source.device;
+        const format = lightingSourcePixelFormat(device);
+
+        const result = new Texture(device, {
+            name: 'lightingSource',
+            cubemap: true,
+            width: 128,
+            height: 128,
+            format: format,
+            type: format === PIXELFORMAT_R8_G8_B8_A8 ? TEXTURETYPE_RGBM : TEXTURETYPE_DEFAULT,
+            addressU: ADDRESS_CLAMP_TO_EDGE,
+            addressV: ADDRESS_CLAMP_TO_EDGE,
+            fixCubemapSeams: fixCubemapSeams,
+            mipmaps: false
+        });
+
+        // copy into top level
+        reprojectTexture(source, result, {
+            numSamples: source.mipmaps ? 1 : 1024
```

Comment:
should 1024 be exposed as a parameter?

Reply:
I think not. We're only resizing a texture here, not convolving or filtering.

We _could_ add the option, but I think would probably confuse users.

Reply:
should it then be based on the resolution of the source?

Reply:
Maybe :) though we're possibly projecting between equirect (or any other projection) and cubemap (and in future visa-versa) making a calculation quite complex.

I've never noticed a need for more samples, but if you feel it's important I can make it an option.

(Though I personally would rather wait till someone finds it necessary before adding).

Reply:
If this can be done on load time / every frame, it'd be useful to have an option to speed it up

Reply:
In dynamic case source should have mipmaps so samples == 1.

Reply:
Maybe just make these a private static members of the class .. so if somebody needs to change them, they're accessible?

Reply:
Actually if we want to support per-frame updating of these then we must have the option of passing in the texture & render target instead of having them allocated each frame.

This will require changes to API.

Reply:
lets support that I think